File: implementations/java/server/src/main/java/com/theshuai/tunnelserver/management/service/HttpMediaPlaybackTicketService.java
package com.theshuai.tunnelserver.management.service;

import com.theshuai.tunnelserver.config.MediaCaptureProperties;
import com.theshuai.tunnelserver.management.model.HttpMediaCapture;
import com.theshuai.tunnelserver.management.repository.HttpMediaCaptureRepository;
import com.theshuai.tunnelserver.management.security.ManagementContext;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.security.SecureRandom;
import java.time.Instant;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class HttpMediaPlaybackTicketService {
    private final HttpMediaCaptureService captureService;
    private final HttpMediaPlaybackService playbackService;
    private final HttpMediaCaptureRepository captureRepository;
    private final MediaCaptureProperties properties;
    private final SecureRandom random = new SecureRandom();
    private final Map<String, Ticket> tickets = new ConcurrentHashMap<>();

    public HttpMediaPlaybackTicketService(HttpMediaCaptureService captureService,
                                          HttpMediaPlaybackService playbackService,
                                          HttpMediaCaptureRepository captureRepository,
                                          MediaCaptureProperties properties) {
        this.captureService = captureService;
        this.playbackService = playbackService;
        this.captureRepository = captureRepository;
        this.properties = properties;
    }

    public PlaybackTicketView create(ManagementContext context, long captureId) {
        return create(context, captureId, false);
    }

    public PlaybackTicketView create(ManagementContext context,
                                     long captureId,
                                     boolean backfillMissing) {
        HttpMediaCapture capture = captureService.requireAccessible(context, captureId);
        if (!HttpMediaCaptureService.STATE_COMPLETE.equals(capture.getState())) {
            throw new IllegalStateException("媒体采集尚未完成");
        }
        if (HttpMediaManifestSupport.MEDIA_SEGMENT.equals(capture.getMediaKind())
                || capture.isInitializationSegment()) {
            throw new IllegalStateException("媒体分段不能独立创建播放会话");
        }
        HttpMediaPlaybackService.PlaybackCacheLayout cacheLayout = null;
        if (!isManifest(capture)) {
            HttpMediaPlaybackService.PlaybackAvailability availability =
                    playbackService.availability(capture);
            if (!availability.playable() && capture.getCapturedBytes() <= 0) {
                throw new IllegalStateException(availability.reason());
            }
            cacheLayout = playbackService.cacheLayout(capture);
        }
        byte[] tokenBytes = new byte[32];
        random.nextBytes(tokenBytes);
        String token = Base64.getUrlEncoder().withoutPadding().encodeToString(tokenBytes);
        Instant expiresAt = Instant.now().plusSeconds(
                Math.max(60, properties.getPlaybackTicketTtlSeconds()));
        tickets.put(token, new Ticket(
                capture.getId(), capture.getTenantId(), expiresAt, backfillMissing));
        String base = "/api/public/media-playback/" + token;
        long totalBytes = cacheLayout == null ? 0 : cacheLayout.totalBytes();
        Long initialRangeStart = null;
        Long initialRangeEnd = null;
        if (cacheLayout != null && capture.getCapturedBytes() > 0 && totalBytes > 0) {
            long start = capture.getContentRangeStart() == null
                    ? 0 : capture.getContentRangeStart();
            long end = capture.getContentRangeEnd() == null
                    ? start + capture.getCapturedBytes() - 1
                    : capture.getContentRangeEnd();
            initialRangeStart = Math.max(0, Math.min(start, totalBytes - 1));
            initialRangeEnd = Math.max(
                    initialRangeStart,
                    Math.min(end, totalBytes - 1));
        }
        return new PlaybackTicketView(
                token,
                capture.getMediaKind(),
                base + "/play",
                base + "/manifest",
                totalBytes,
                initialRangeStart,
                initialRangeEnd,
                cacheLayout == null ? List.of() : cacheLayout.cachedRanges(),
                backfillMissing,
                expiresAt.toString());
    }

    public ResolvedTicket resolve(String token) {
        Ticket ticket = tickets.get(token);
        if (ticket == null || ticket.expiresAt().isBefore(Instant.now())) {
            if (ticket != null) {
                tickets.remove(token, ticket);
            }
            throw new IllegalArgumentException("媒体播放票据无效或已过期");
        }
        HttpMediaCapture capture = captureRepository
                .findByIdAndTenantId(ticket.captureId(), ticket.tenantId())
                .orElseThrow(() -> new IllegalArgumentException("媒体采集记录不存在"));
        return new ResolvedTicket(
                token, capture, ticket.expiresAt(), ticket.backfillMissing());
    }

    @Scheduled(fixedDelay = 60_000L)
    public void cleanup() {
        Instant now = Instant.now();
        tickets.entrySet().removeIf(entry -> entry.getValue().expiresAt().isBefore(now));
    }

    private record Ticket(
            long captureId,
            String tenantId,
            Instant expiresAt,
            boolean backfillMissing
    ) {
    }

    public record ResolvedTicket(
            String token,
            HttpMediaCapture capture,
            Instant expiresAt,
            boolean backfillMissing
    ) {
        public String assetBasePath() {
            return "/api/public/media-playback/" + token + "/asset";
        }
    }

    public record PlaybackTicketView(
            String ticket,
            String mediaKind,
            String playUrl,
            String manifestUrl,
            long totalBytes,
            Long initialRangeStart,
            Long initialRangeEnd,
            List<HttpMediaPlaybackService.PlaybackByteRange> cachedRanges,
            boolean backfillMissing,
            String expiresAt
    ) {
    }

    private boolean isManifest(HttpMediaCapture capture) {
        return HttpMediaManifestSupport.HLS_MANIFEST.equals(capture.getMediaKind())
                || HttpMediaManifestSupport.DASH_MANIFEST.equals(capture.getMediaKind());
    }
}

File: implementations/csharp/server/src/Specus.Server/Management/HttpMediaPlaybackTicketService.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;
using Microsoft.Extensions.Options;
using Specus.Server.Configuration;
using Specus.Server.Data.Entities;

namespace Specus.Server.Management;

public sealed class HttpMediaPlaybackTicketService : BackgroundService
{
    private readonly ConcurrentDictionary<string, Ticket> _tickets = new(StringComparer.Ordinal);
    private readonly MediaCaptureOptions _options;
    private readonly TimeProvider _timeProvider;

    public HttpMediaPlaybackTicketService(IOptions<MediaCaptureOptions> options,
        TimeProvider timeProvider)
        : this(options.Value, timeProvider)
    {
    }

    internal HttpMediaPlaybackTicketService(MediaCaptureOptions options, TimeProvider timeProvider)
    {
        _options = options;
        _timeProvider = timeProvider;
    }

    public async Task<HttpMediaPlaybackTicketView> CreateAsync(HttpMediaCapture capture,
        HttpMediaPlaybackService playbackService, bool backfillMissing,
        CancellationToken cancellationToken)
    {
        if (capture.State != HttpMediaCaptureService.StateComplete)
        {
            throw new InvalidOperationException("媒体采集尚未完成");
        }
        if (capture.MediaKind == HttpMediaManifestSupport.MediaSegment
            || capture.InitializationSegment)
        {
            throw new InvalidOperationException("媒体分段不能独立创建播放会话");
        }
        HttpMediaPlaybackService.PlaybackCacheLayout? layout = null;
        if (!HttpMediaManifestSupport.IsManifest(capture.MediaKind))
        {
            var availability = await playbackService.AvailabilityAsync(capture, cancellationToken)
                .ConfigureAwait(false);
            if (!availability.Playable && capture.CapturedBytes <= 0)
            {
                throw new InvalidOperationException(availability.Reason);
            }
            layout = await playbackService.CacheLayoutAsync(capture, cancellationToken)
                .ConfigureAwait(false);
        }
        var token = Base64Url(RandomNumberGenerator.GetBytes(32));
        var expiresAt = _timeProvider.GetUtcNow()
            .AddSeconds(Math.Max(60, _options.PlaybackTicketTtlSeconds));
        _tickets[token] = new Ticket(capture.Id, capture.TenantId, expiresAt, backfillMissing);
        var basePath = "/api/public/media-playback/" + token;
        var totalBytes = layout?.TotalBytes ?? 0;
        long? initialStart = null;
        long? initialEnd = null;
        if (layout is not null && capture.CapturedBytes > 0 && totalBytes > 0)
        {
            var start = capture.ContentRangeStart ?? 0;
            var end = capture.ContentRangeEnd ?? start + capture.CapturedBytes - 1;
            initialStart = Math.Max(0, Math.Min(start, totalBytes - 1));
            initialEnd = Math.Max(initialStart.Value, Math.Min(end, totalBytes - 1));
        }
        return new HttpMediaPlaybackTicketView(token, capture.MediaKind,
            basePath + "/play", basePath + "/manifest", totalBytes, initialStart, initialEnd,
            layout?.CachedRanges ?? [], backfillMissing, expiresAt.ToString("O"));
    }

    public ResolvedTicket Resolve(string token)
    {
        Ticket? ticket = null;
        if (string.IsNullOrWhiteSpace(token) || !_tickets.TryGetValue(token, out ticket)
            || ticket.ExpiresAt < _timeProvider.GetUtcNow())
        {
            if (ticket is not null)
            {
                _tickets.TryRemove(new KeyValuePair<string, Ticket>(token, ticket));
            }
            throw new ArgumentException("媒体播放票据无效或已过期");
        }
        return new ResolvedTicket(token, ticket.CaptureId, ticket.TenantId, ticket.ExpiresAt,
            ticket.BackfillMissing);
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        using var timer = new PeriodicTimer(TimeSpan.FromMinutes(1));
        while (await timer.WaitForNextTickAsync(stoppingToken).ConfigureAwait(false))
        {
            var now = _timeProvider.GetUtcNow();
            foreach (var entry in _tickets)
            {
                if (entry.Value.ExpiresAt < now)
                {
                    _tickets.TryRemove(entry);
                }
            }
        }
    }

    private static string Base64Url(byte[] bytes) => Convert.ToBase64String(bytes)
        .TrimEnd('=').Replace('+', '-').Replace('/', '_');

    private sealed record Ticket(long CaptureId, string TenantId, DateTimeOffset ExpiresAt,
        bool BackfillMissing);

    public sealed record ResolvedTicket(string Token, long CaptureId, string TenantId,
        DateTimeOffset ExpiresAt, bool BackfillMissing)
    {
        public string AssetBasePath => "/api/public/media-playback/" + Token + "/asset";
    }
}
